File: frontend/types/litigation.ts
// Mirrors backend/app/schemas/litigation.py — Phase 9.3 Litigation & Case Intelligence.

export type PartyType = "individual" | "organization" | "unknown";
export type ProceduralRole = "plaintiff" | "defendant" | "third_party" | "applicant" | "respondent" | "unknown";

export interface CaseParty {
  id: string;
  case_id: string;
  name: string;
  party_type: PartyType;
  procedural_role: ProceduralRole;
  identifiers: Record<string, unknown>;
}

export type CaseDocumentRole =
  | "contract" | "addendum" | "invoice" | "act" | "correspondence" | "claim" | "response"
  | "court_filing" | "court_decision" | "expert_report" | "payment_document" | "other";

export interface CaseDocument {
  id: string;
  case_id: string;
  document_id: string;
  role: CaseDocumentRole;
  document_title: string;
  document_status: string;
}

export type FactType = "date" | "amount" | "party" | "event" | "other";
export type FactStatus = "asserted" | "supported" | "disputed" | "contradicted" | "inferred" | "unknown";

export interface CaseFactEvidence {
  document_id: string;
  document_title: string;
  chunk_id: string | null;
  page_number: number | null;
  section_path: string | null;
  excerpt: string;
}

export interface CaseFact {
  id: string;
  case_id: string;
  statement: string;
  fact_type: FactType;
  status: FactStatus;
  normalized_value: string | null;
  evidence: CaseFactEvidence[];
  created_at: string | null;
}

export type DateType = "exact" | "calculated" | "approximate" | "unknown";

export interface CaseEvent {
  id: string;
  case_id: string;
  event_date: string | null;
  date_type: DateType;
  description: string;
  event_type: string | null;
  source_fact_id: string | null;
}

export type ContradictionType = "date_mismatch" | "amount_mismatch" | "party_mismatch" | "other";

export interface CaseContradiction {
  id: string;
  case_id: string;
  contradiction_type: ContradictionType;
  description: string;
  fact_a_id: string;
  fact_a_statement: string;
  fact_b_id: string;
  fact_b_statement: string;
}

export type EvidenceStrength = "strong" | "moderate" | "weak" | "conflicted" | "insufficient";

export interface EvidenceMatrixRow {
  fact_statement: string;
  fact_type: FactType;
  normalized_value: string;
  strength: EvidenceStrength;
  reasons: string[];
  corroboration_count: number;
}

export interface CaseAnalysisSummary {
  case_id: string;
  fact_count: number;
  contradiction_count: number;
  event_count: number;
}

// --- Case Result Summary (client-facing) ---

export interface MoneyFlowTransaction {
  payment_order_id: string;
  document_id: string;
  payment_date: string | null;
  amount: string | null;
  payer: string | null;
  recipient: string | null;
  referenced_contract_date: string | null;
}

export interface MoneyFlow {
  transaction_count: number;
  transactions: MoneyFlowTransaction[];
  total_amount: string;
  referenced_contract_dates: Record<string, number>;
  referenced_contract_numbers: Record<string, number>;
}

export interface CaseSnapshot {
  party_names: string[];
  document_count: number;
  payment_count: number;
  total_amount: string;
  key_dates: [string | null, string][];
}

export interface KeyFinding {
  severity: string;
  statement: string;
  source_document_id: string;
  source_document_title: string;
  page_number: number | null;
  excerpt: string;
  confidence: string;
  caveat: string | null;
}

export interface MissingEvidenceItem {
  priority: string;
  description: string;
  why_it_matters: string;
  source_document_id: string | null;
  source_document_title: string | null;
}

export interface NextBestAction {
  priority: number;
  action: string;
  why: string;
}

export type RelationshipType = "director" | "shareholder" | "member" | "other";
export type RelationshipVerificationStatus = "unverified" | "document_supported" | "externally_verified" | "conflicting";

export interface PartyRelationshipFinding {
  subject_name: string;
  related_party_name: string;
  relationship_type: RelationshipType;
  relationship_start: string | null;
  relationship_end: string | null;
  timing_note: string;
  why_it_may_matter: string;
  what_is_still_needed: string[];
  verification_status: RelationshipVerificationStatus;
  source_document_id: string | null;
  source_document_title: string | null;
  source_excerpt: string | null;
}

export interface CaseResultSummary {
  case_snapshot: CaseSnapshot;
  key_findings: KeyFinding[];
  money_flow: MoneyFlow;
  what_this_may_mean: string[];
  missing_critical_evidence: MissingEvidenceItem[];
  next_best_actions: NextBestAction[];
  legal_kb_warning: string | null;
  party_relationship_findings: PartyRelationshipFinding[];
}

// --- Case Intelligence: party relationships, hypothesis register, related litigation ---

export type HypothesisCategory = "fact" | "counsel_hypothesis" | "ai_inference" | "missing_evidence";

export interface CasePartyRelationship {
  id: string;
  case_id: string;
  subject_party_id: string;
  related_party_id: string;
  relationship_type: RelationshipType;
  ownership_percentage: string | null;
  start_date: string | null;
  end_date: string | null;
  source_document_id: string | null;
  source_excerpt: string | null;
  verification_status: RelationshipVerificationStatus;
  notes: string | null;
}

export interface CaseHypothesis {
  id: string;
  case_id: string;
  category: HypothesisCategory;
  statement: string;
  required_verification: string[];
  related_relationship_id: string | null;
  source: string | null;
}

export interface CaseRelatedLitigation {
  id: string;
  case_id: string;
  court: string | null;
  case_number: string | null;
  parties_description: string | null;
  subject_matter: string | null;
  amount_in_dispute: string | null;
  status: string | null;
  note: string | null;
  contextual_note: string;
}

// --- Master Case Report ---

export type FindingCategory =
  | "claim_contradiction" | "payment_pattern" | "contract_formation" | "contract_mismatch" | "course_of_dealing"
  | "party_conduct" | "interest_calculation" | "procedural" | "corporate_relationship" | "related_litigation"
  | "evidence_gap" | "legal_argument" | "risk" | "other";

export interface MasterFinding {
  id: string;
  category: FindingCategory;
  title: string;
  statement: string;
  supporting_facts: string[];
  contradicting_facts: string[];
  source_document_ids: string[];
  source_document_titles: string[];
  excerpts: string[];
  page_numbers: (number | null)[];
  helps_side: string;
  hurts_side: string;
  strength: string;
  confidence: string;
  legal_significance: string;
  counterargument: string | null;
  response_to_counterargument: string | null;
  caveat: string | null;
  missing_evidence: string[];
  recommended_action: string | null;
  verification_status: string;
  alternative_explanations: string[];
  what_would_strengthen: string[];
  what_would_weaken: string[];
  legal_research_required: boolean;
}

export interface CaseOnePager {
  case_position: string;
  strongest_point: string | null;
  biggest_risk: string | null;
  money_at_stake: string;
  top_arguments: string[];
  top_risks: string[];
  what_opponent_must_explain: string[];
  what_court_likely_focuses_on: string | null;
  missing_p0_evidence: string[];
  next_best_action: string | null;
}

export interface CourtScenario {
  scenario: string;
  why_court_could_get_there: string;
  facts_supporting: string[];
  facts_against: string[];
  label: string;
}

export interface DraftResponseSection {
  section: string;
  argument: string;
  supporting_finding_ids: string[];
  caution: string | null;
}

export interface BurdenItem {
  proposition: string;
  side: string;
  current_evidence: string[];
  contrary_evidence: string[];
  status: string;
  weakness: string | null;
  how_to_attack: string | null;
}

export interface CaseMap {
  claimed_amounts: string[];
  claim_dates: string[];
  note: string;
}

export interface ContractVersionTerms {
  document_id: string;
  document_title: string;
  amounts: string[];
  interest_rate: string | null;
  maturity_dates: string[];
  formation_clause_present: boolean;
  signature_status: string;
}

export interface MasterCaseReport {
  one_pager: CaseOnePager;
  case_map: CaseMap;
  findings: MasterFinding[];
  burden_map: BurdenItem[];
  court_scenarios: CourtScenario[];
  opposing_party_questions: string[];
  draft_response_structure: DraftResponseSection[];
  contract_version_matrix: ContractVersionTerms[];
  money_flow: MoneyFlow;
  legal_kb_warning: string | null;
}

File: backend/app/schemas/litigation.py
from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict

from app.domains.litigation.master_report import FindingCategory
from app.models.matters import (
    AllegationType,
    CaseDocumentRole,
    ContradictionType,
    DateType,
    FactStatus,
    FactType,
    HypothesisCategory,
    PartyType,
    PaymentExecutionStatus,
    ProceduralRole,
    RelationshipType,
    RelationshipVerificationStatus,
)


class CasePartyCreate(BaseModel):
    name: str
    party_type: PartyType = PartyType.UNKNOWN
    procedural_role: ProceduralRole = ProceduralRole.UNKNOWN
    identifiers: dict = {}


class CasePartyOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    case_id: uuid.UUID
    name: str
    party_type: PartyType
    procedural_role: ProceduralRole
    identifiers: dict


class CaseDocumentAttach(BaseModel):
    document_id: uuid.UUID
    role: CaseDocumentRole = CaseDocumentRole.OTHER


class CaseDocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    case_id: uuid.UUID
    document_id: uuid.UUID
    role: CaseDocumentRole
    document_title: str
    document_status: str


class CaseFactEvidenceOut(BaseModel):
    document_id: uuid.UUID
    document_title: str
    chunk_id: uuid.UUID | None
    page_number: int | None
    section_path: str | None
    excerpt: str


class CaseFactOut(BaseModel):
    id: uuid.UUID
    case_id: uuid.UUID
    statement: str
    fact_type: FactType
    status: FactStatus
    normalized_value: str | None
    evidence: list[CaseFactEvidenceOut]
    created_at: datetime | None = None


class CaseEventOut(BaseModel):
    id: uuid.UUID
    case_id: uuid.UUID
    event_date: date | None
    date_type: DateType
    description: str
    event_type: str | None
    source_fact_id: uuid.UUID | None


class CaseContradictionOut(BaseModel):
    id: uuid.UUID
    case_id: uuid.UUID
    contradiction_type: ContradictionType
    description: str
    fact_a_id: uuid.UUID
    fact_a_statement: str
    fact_b_id: uuid.UUID
    fact_b_statement: str


class EvidenceMatrixRowOut(BaseModel):
    fact_statement: str
    fact_type: FactType
    normalized_value: str
    strength: str
    reasons: list[str]
    corroboration_count: int


# --- E1: case allegations ---


class CaseAllegationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    case_id: uuid.UUID
    document_id: uuid.UUID
    document_title: str
    chunk_id: uuid.UUID | None
    page_number: int | None
    statement_text: str
    excerpt: str
    allegation_type: AllegationType
    created_at: datetime | None = None


# --- E3: structured payment orders ---


class CasePaymentOrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    case_id: uuid.UUID
    document_id: uuid.UUID
    document_title: str
    page_number: int | None
    payment_date: date | None
    amount: str | None
    payer: str | None
    recipient: str | None
    payment_purpose: str | None
    referenced_contract_type: str | None
    referenced_contract_date: date | None
    referenced_contract_number: str | None
    execution_status: PaymentExecutionStatus
    excerpt: str


class MoneyFlowTransactionOut(BaseModel):
    payment_order_id: uuid.UUID
    document_id: uuid.UUID
    payment_date: date | None
    amount: str | None
    payer: str | None
    recipient: str | None
    referenced_contract_date: date | None


class MoneyFlowOut(BaseModel):
    transaction_count: int
    transactions: list[MoneyFlowTransactionOut]
    total_amount: str
    referenced_contract_dates: dict[str, int]
    referenced_contract_numbers: dict[str, int]


# --- E2: CLAIM_VS_EVIDENCE (computed, never persisted) ---


class ClaimEvidenceContradictionOut(BaseModel):
    contradiction_type: ContradictionType
    allegation_id: uuid.UUID
    allegation_document_id: uuid.UUID
    allegation_document_title: str
    allegation_page: int | None
    allegation_excerpt: str
    evidence_id: uuid.UUID
    evidence_document_id: uuid.UUID
    evidence_document_title: str
    evidence_page: int | None
    evidence_excerpt: str
    referenced_contract_date: date | None
    reason: str
    caveat: str
    confidence: str


# --- Case Result Summary (client-facing, template/deterministic synthesis) ---


class CaseSnapshotOut(BaseModel):
    party_names: list[str]
    document_count: int
    payment_count: int
    total_amount: str
    key_dates: list[tuple[date | None, str]]


class KeyFindingOut(BaseModel):
    severity: str
    statement: str
    source_document_id: uuid.UUID
    source_document_title: str
    page_number: int | None
    excerpt: str
    confidence: str
    caveat: str | None


class MissingEvidenceItemOut(BaseModel):
    priority: str
    description: str
    why_it_matters: str
    source_document_id: uuid.UUID | None = None
    source_document_title: str | None = None


class NextBestActionOut(BaseModel):
    priority: int
    action: str
    why: str


class PartyRelationshipFindingOut(BaseModel):
    subject_name: str
    related_party_name: str
    relationship_type: RelationshipType
    relationship_start: date | None
    relationship_end: date | None
    timing_note: str
    why_it_may_matter: str
    what_is_still_needed: list[str]
    verification_status: RelationshipVerificationStatus
    source_document_id: uuid.UUID | None
    source_document_title: str | None
    source_excerpt: str | None


class CaseResultSummaryOut(BaseModel):
    case_snapshot: CaseSnapshotOut
    key_findings: list[KeyFindingOut]
    money_flow: MoneyFlowOut
    what_this_may_mean: list[str]
    missing_critical_evidence: list[MissingEvidenceItemOut]
    next_best_actions: list[NextBestActionOut]
    legal_kb_warning: str | None
    party_relationship_findings: list[PartyRelationshipFindingOut] = []


# --- Case Intelligence: party relationships, hypothesis register, related litigation ---


class CasePartyRelationshipCreate(BaseModel):
    subject_party_id: uuid.UUID
    related_party_id: uuid.UUID
    relationship_type: RelationshipType
    ownership_percentage: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    source_document_id: uuid.UUID | None = None
    source_excerpt: str | None = None
    verification_status: RelationshipVerificationStatus = RelationshipVerificationStatus.UNVERIFIED
    notes: str | None = None


class CasePartyRelationshipOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    case_id: uuid.UUID
    subject_party_id: uuid.UUID
    related_party_id: uuid.UUID
    relationship_type: RelationshipType
    ownership_percentage: str | None
    start_date: date | None
    end_date: date | None
    source_document_id: uuid.UUID | None
    source_excerpt: str | None
    verification_status: RelationshipVerificationStatus
    notes: str | None


class CaseHypothesisCreate(BaseModel):
    category: HypothesisCategory
    statement: str
    required_verification: list[str] = []
    related_relationship_id: uuid.UUID | None = None
    source: str | None = None


class CaseHypothesisOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    case_id: uuid.UUID
    category: HypothesisCategory
    statement: str
    required_verification: list[str]
    related_relationship_id: uuid.UUID | None
    source: str | None


class CaseRelatedLitigationCreate(BaseModel):
    court: str | None = None
    case_number: str | None = None
    parties_description: str | None = None
    subject_matter: str | None = None
    amount_in_dispute: str | None = None
    status: str | None = None
    note: str | None = None


class CaseRelatedLitigationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    case_id: uuid.UUID
    court: str | None
    case_number: str | None
    parties_description: str | None
    subject_matter: str | None
    amount_in_dispute: str | None
    status: str | None
    note: str | None
    # Computed at read time, never stored verbatim as a causal claim — see
    # case_relationships.py's build_related_litigation_note().
    contextual_note: str


# --- Master Case Report ---


class MasterFindingOut(BaseModel):
    id: str
    category: FindingCategory
    title: str
    statement: str
    supporting_facts: list[str]
    contradicting_facts: list[str]
    source_document_ids: list[uuid.UUID]
    source_document_titles: list[str]
    excerpts: list[str]
    page_numbers: list[int | None]
    helps_side: str
    hurts_side: str
    strength: str
    confidence: str
    legal_significance: str
    counterargument: str | None
    response_to_counterargument: str | None
    caveat: str | None
    missing_evidence: list[str]
    recommended_action: str | None
    verification_status: str


class CaseOnePagerOut(BaseModel):
    case_position: str
    strongest_point: str | None
    biggest_risk: str | None
    money_at_stake: str
    top_arguments: list[str]
    top_risks: list[str]
    what_opponent_must_explain: list[str]
    what_court_likely_focuses_on: str | None
    missing_p0_evidence: list[str]
    next_best_action: str | None


class CourtScenarioOut(BaseModel):
    scenario: str
    why_court_could_get_there: str
    facts_supporting: list[str]
    facts_against: list[str]
    label: str


class DraftResponseSectionOut(BaseModel):
    section: str
    argument: str
    supporting_finding_ids: list[str]
    caution: str | None


class BurdenItemOut(BaseModel):
    proposition: str
    side: str
    current_evidence: list[str]
    contrary_evidence: list[str]
    status: str
    weakness: str | None
    how_to_attack: str | None


class CaseMapOut(BaseModel):
    claimed_amounts: list[str]
    claim_dates: list[str]
    note: str


class ContractVersionTermsOut(BaseModel):
    document_id: uuid.UUID
    document_title: str
    amounts: list[str]
    interest_rate: str | None
    maturity_dates: list[str]
    formation_clause_present: bool
    signature_status: str


class MasterCaseReportOut(BaseModel):
    one_pager: CaseOnePagerOut
    case_map: CaseMapOut
    findings: list[MasterFindingOut]
    burden_map: list[BurdenItemOut]
    court_scenarios: list[CourtScenarioOut]
    opposing_party_questions: list[str]
    draft_response_structure: list[DraftResponseSectionOut]
    contract_version_matrix: list[ContractVersionTermsOut]
    money_flow: MoneyFlowOut
    legal_kb_warning: str | None
